lib/stdlib/src/shell_docs.erl

```diff
@@ -0,0 +1,672 @@
+                          chunk_elements()} | binary().
+-type chunk_elements() :: [chunk_element()].
+-type docs_v1() :: #docs_v1{}.
+
+
+-spec validate(Module) -> ok when
+      Module :: module() | docs_v1().
+%% Simple validation of erlang doc chunk. Check that all tags are supported and
+%% that the signature is correct.
+validate(Module) when is_atom(Module) ->
+    {ok, Doc} = code:get_doc(Module),
+    validate(Doc);
+validate(#docs_v1{ module_doc = MDocs, docs = AllDocs }) ->
+    _ = maps:map(fun(_Key,MDoc) -> validate(binary_to_term(MDoc)) end, MDocs),
+    lists:map(fun({_,_Anno, Sig, Docs, _Meta}) ->
+                      case lists:all(fun erlang:is_binary/1, Sig) of
+                          false -> throw({invalid_signature,Sig});
+                          true -> ok
+                      end,
+                      maps:map(fun(_Key,Doc) -> validate(binary_to_term(Doc)) end, Docs)
+              end, AllDocs);
+validate([H|T]) when is_tuple(H) ->
+    _ = validate(H),
+    validate(T);
+validate({Tag,Attr,Content}) ->
+    case lists:member(Tag,?ALL_TAGS) of
+        false ->
+            throw({invalid_tag,Tag});
+        true ->
+            ok
+    end,
+    true = is_list(Attr),
+    validate(Content);
+validate([Chars | T]) when is_binary(Chars) ->
+    validate(T);
+validate([]) ->
+    ok.
+
+%% Follows algorithm described here:
+%% * https://medium.com/@patrickbrosset/when-does-white-space-matter-in-html-b90e8a7cdd33
+%% which in turn follows this:
+%% * https://www.w3.org/TR/css-text-3/#white-space-processing
+-spec normalize(Docs) -> NormalizedDocs when
+      Docs :: chunk_elements(),
+      NormalizedDocs :: chunk_elements().
+normalize(Docs) ->
+    Trimmed = normalize_trim(Docs,true),
+    normalize_space(Trimmed).
+
+normalize_trim(Bin,true) when is_binary(Bin) ->
+    %% Remove any whitespace (except \n) before or after a newline
+    NoSpace = re:replace(Bin,"[^\\S\n]*\n+[^\\S\n]*","\n",[global]),
+    %% Replace any tabs with space
+    NoTab = re:replace(NoSpace,"\t"," ",[global]),
+    %% Replace any newlines with space
+    NoNewLine = re:replace(NoTab,"\\v"," ",[global]),
+    %% Replace any sequences of \s with a single " "
+    re:replace(NoNewLine,"\\s+"," ",[global,{return,binary}]);
+normalize_trim(Bin,false) when is_binary(Bin) ->
+    Bin;
+normalize_trim([{Tag,Attr,Content}|T],Trim) when Tag =:= pre;
+                                                 Tag =:= code ->
+    [{Tag,Attr,normalize_trim(Content,false)} | normalize_trim(T,Trim)];
+normalize_trim([{Tag,Attr,Content}|T],Trim) ->
+    [{Tag,Attr,normalize_trim(Content,Trim)} | normalize_trim(T,Trim)];
+normalize_trim([<<>>|T],Trim) ->
+    normalize_trim(T,Trim);
+normalize_trim([B1,B2|T],Trim) when is_binary(B1),is_binary(B2) ->
+    normalize_trim([<<B1/binary,B2/binary>> | T],Trim);
+normalize_trim([H|T],Trim) ->
+    [normalize_trim(H,Trim) | normalize_trim(T,Trim)];
+normalize_trim([],_Trim) ->
+    [].
```

Comment:
Should we do this when writing to the chunk? 🤔 Otherwise every tool processing the chunk will have to do it.

Reply:
Please ignore. I see this is already used by docgen. :D

Reply:
My plan is to call it both when generating and when rendering. That way tools that generate can skip it if they want and renderers that want to render Erlang/OTP docs dont have to normalize.

Reply:
Fantastic!